feature_engine/preprocessing/match_categories.py

```diff
@@ -0,0 +1,159 @@
+import warnings
+from typing import List, Optional, Union
+
+import pandas as pd
+from feature_engine._docstrings.fit_attributes import (
+    _feature_names_in_docstring,
+    _n_features_in_docstring,
+    _variables_attribute_docstring,
+)
+from feature_engine._docstrings.substitute import Substitution
+from feature_engine.dataframe_checks import _check_contains_na, check_X
+from feature_engine.encoding._docstrings import (
+    _errors_docstring,
+    _ignore_format_docstring,
+    _variables_docstring,
+)
+from feature_engine.encoding._helper_functions import check_parameter_errors
+from feature_engine.encoding.base_encoder import (
+    CategoricalInitMixin,
+    CategoricalMethodsMixin,
+)
+
+
+@Substitution(
+    ignore_format=_ignore_format_docstring,
+    errors=_errors_docstring,
+    variables=_variables_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+)
+class MatchCategories(CategoricalInitMixin, CategoricalMethodsMixin):
+    """
+    MatchCategories() ensures that categorical variables are encoded as pandas
+    `'categorical'` dtype, instead of generic python `'object'` or other dtypes.
+
+    Under the hood, `'categorical'` dtype is a representation that maps each
+    category to an integer, thus providing a more memory-efficient object
+    structure than e.g., 'str', and allowing faster grouping, mapping, and similar
+    operations on the resulting object.
+
+    MatchCategories() remembers the encodings or levels that represent each
+    category, and can thus can be used to ensure that the correct encoding gets
+    applied when passing categorical data to modeling packages that support this
+    dtype, or to prevent unseen categories from reaching a further transformer
+    or estimator in a pipeline, for example.
+
+    Parameters
+    ----------
+    {variables}
+
+    {ignore_format}
+
+    {errors}
+
+    Attributes
+    ----------
+    category_dict_:
+        Dictionary with the category encodings assigned to each variable.
+
+    {variables_}
+
+    {feature_names_in_}
+
+    {n_features_in_}
+
+    Methods
+    -------
+    fit:
+        Learn the encodings or levels to use for each variable.
+
+    transform:
+        Enforce the type of categorical variables as dtype `categorical`.
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+        ignore_format: bool = False,
+        errors: str = "raise",
+    ) -> None:
+        check_parameter_errors(errors, ["ignore", "raise"])
+        super().__init__(variables, ignore_format, allow_missing=errors != "raise")
```

Comment:
the logic for missing values needs to be moved here.

Also, we normally call this parameter "missing" and takes values "raise" and "ignore". So for consistency, could we use the same names?

Reply:
In the other categorical encoders (such as `CountFrequency`), it it called "errors".